### DISCUSSION TITLE
recoil (with selector?) 로 api 요청을 모두 대체할 순 없을까!?

### DISCUSSION THREAD

**Discussion by sbjang123456** [sbjang123456/TIL#119] [General]
redux-saga, swr, react-query 와 같은 서드파티 라이브러리들은 서버 API 의 `GET`, `POST`, `DELETE`, `PUT`, `PATCH` 등 모두 요청할 수 있도록 되어있다.
그러나 아직까지 recoil 공식 문서에 따르면 `selector`를 통한 비동기 `GET` 요청은 가이드 되어있지만, POST 와 같은 요청들에 대한 명확한 해답은 찾아볼 수 없었다. 
어떻게 하면 전역상태관리 recoil 라이브러리 하나로 비동기 요청까지 퉁칠 수 있을까?ㅎㅎ
> 개인적으로 나는 전역상태관리, 서비스요청, 컴포넌트, 컨테이너와 같이 로직 분리를 상당히 좋아한다..
#122

**Comment by SooJungChae** [comment]
위메포에서 atom 으로 POST/PUT 같은 요청해본적이 있었는데 이렇게 해도 되나 ...? 하는 기억이 있네요 ㅋㅋㅋ
지금보니까 recoil 은 서버에서 'GET' 해온 데이터의 상태를 관리하기 위한 라이브러리 같아요.

**recoil**
> For reasons of compatibility and simplicity, it's best to use React's built-in state management capabilities rather than external global state. But React has certain limitations:
> Component state can only be shared by pushing it up to the common ancestor, but this might include a huge tree that then needs to re-render.
Context can only store a single value, not an indefinite set of values each with its own consumers.
Both of these make it difficult to code-split the top of the tree (where the state has to live) from the leaves of the tree (where the state is used).

**react-query**
> React Query is often described as the missing data-fetching library for React, but in more technical terms, it makes fetching, caching, synchronizing and updating server state in your React applications a breeze.

recoil 은 서버 데이터 핸들링 보다는 내부 상태관리에 초점을 맞추고 있는 것 같아요. 그래서 찾기 어려운게 아닐까요??

+ react-query 에서 보시면 'useQuery' 와 'useMutation' 이라는 용어를 사용하고 있어요.
'GET' 을 할때는 'useQuery' 를 사용하고 'CREATE' 나 'UPDATE' 같은 db 데이터를 바꿀 땐 'useMutation' 을 사용하도록 권장합니다.

때마침(?) recoil 에서는 'query' 용어를 사용하고 있으니 '내부 상태관리에 필요한 라이브러리이다'에 더 힘을 실어주는 것 같아보이네요~

고로! 개인적으로는 recoil 은 ContextApi 를 잘 사용하려고 도와주는 라이브러리 같고, react-query 는 서버 통신해서 데이터 핸들링을 하는 걸 도와주는 라이브러리 같습니다 ㅋㅋㅋ

recoil 하나로 API method 처리를 다 만들어 보고 싶으시다는 것이었는데 말이 길었군요 허허. 
결론은 react-query 를 사용하시는 걸 추천드리는데 어떻게 처리하실지 궁금하기도 하네요~
